### PR TITLE
Experiment: Upsell screen in Jetpack purchase flow

### PR DESCRIPTION
**Pull request by monsieur-z**
### Changes proposed in this Pull Request

This PR implements a new screen to upsell Jetpack products, as an A/B test. This new screen will appear:
- Between the pricing page and the checkout page
- Only if a user selects Backup, Scan, or Antispam (the upsold product is Security)

Context: pbNhbs-2pq-p2

This PR implements all the logic for this experiment and new page. The design of the page is handled in #64329.

### Implementation notes

- This new page has a dedicated URL, depending on where the user is seeing the landing page from:
  - `cloud.jetpack.com/pricing/upsell/:product/:site?`
  - `wordpress.com/pricing/upsell/:product/:site?`
  - `wordpress.com/jetpack/connect/store/upsell/:product`
  - `wordpress.com/jetpack/connect/plans/upsell/:product/:site`
- Name of the experiment: `calypso_jetpack_upsell_page_2022_06 `

### Testing instructions

_Note: clicking the close button in the checkout page should redirect you to the pricing page (passed in the `checkoutBackUrl` query param). This doesn't work as expected, and should be addressed in a separate PR._

#### Prerequisites

- Make sure you have a site connected to Jetpack with no subscriptions
- Check for JS errors in the console

#### Regression testing

- If you've tested this PR before, make sure to delete the `explat-experiment--jetpack_upsell_page_2022_05` key in your local storage, or assign yourself to the control variation in the experiment tool.
- In Calypso blue, visit `/plans/:site`
- Select a product, and make sure you're directed to the checkout page with the product in the cart
- Repeat the same operation with a couple more products
- Visit `/plans/upsell/jetpack_backup_t1_yearly/:site`: you should be redirected to the pricing page (you might briefly see the upsell page flash, but that's acceptable since we'll eventually keep or get rid of the page definitively)

#### Upsell screen

- Assign yourself to the treatment variation
- Calypso pricing page: 
  - Visit `/plans/:site`
  - Check that selecting Backup, Scan, or Anti-spam in the pricing page leads you to the upsell page
  - In the upsell page:
    - Check that the copy makes sense
    - Check that the price difference is accurate
    - Check that both links do what they advertise
    - Refresh the page, and verify that you see the same information
  - Check that selecting another product (than the 3 mentioned above) in the pricing page leads you directly to the checkout page
  - Visit `/plans/upsell/jetpack_complete/:site`, and check that you're redirected to the checkout page
- Calypso connect page: 
  - Visit `/jetpack/connect/store`
  - Repeat the aforementioned steps
  - Visit `/jetpack/connect/plans/:site`
  - Repeat the same steps
- Cloud pricing page:
  - Visit `/pricing`
  - Repeat the aforementioned steps
  - Visit `/pricing/:site`
  - Repeat the same steps

### CODE REVIEW

File: client/jetpack-cloud/sections/pricing/index.ts [hunk 1]
```diff
@@ -14,6 +16,12 @@ export default function (): void {
 	page( '/plans/storage/:site', ( { params } ) =>
 		page.redirect( `/pricing/storage/${ params.site }` )
 	);
+	page( '/:locale/plans/upsell/:product/:site?', ( { params } ) =>
```
Comment (monsieur-z) [inline client/jetpack-cloud/sections/pricing/index.ts:19]: Mimic existing redirects.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+			};
+		}
+
+		return {};
+	}, [ upsellSlug ] );
+
+	const productName = productItem?.displayName;
+	const upsellName = upsellItem?.displayName;
+
+	const productPriceObj = useItemPrice( siteId, productItem );
```
Comment (monsieur-z) [inline client/my-sites/plans/jetpack-plans/upsell/index.tsx:101]: Here, we use the same price selector used in the pricing page.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	const upsellPriceObj = useItemPrice( siteId, upsellItem );
+	const priceDelta =
+		( upsellPriceObj?.discountedPrice || upsellPriceObj?.originalPrice ) -
+		( productPriceObj?.discountedPrice || productPriceObj?.originalPrice );
+	const isLoadingPrice = productPriceObj?.isFetching || upsellPriceObj?.isFetching;
+
+	useEffect( () => {
```
Comment (monsieur-z) [inline client/my-sites/plans/jetpack-plans/upsell/index.tsx:108]: Redirect to checkout page if:
- User is not assigned the treatment variation
- The product to upsell is not a valid product to upsell
- Site has already a subscription to the upsold product

**Comment by monsieur-z** [comment]
@Automattic/jetpack-voyager, do you foresee any issue with the changes made to `buildCheckoutURL`?

**Comment by mattgawarecki** [comment]
> @Automattic/jetpack-voyager, do you foresee any issue with the changes made to `buildCheckoutURL`?

Considering the added function parameters are optional, and any new behavior wouldn't execute unless their default values are overridden, I don't expect any regression issues with the `buildCheckoutURL` function specifically. That said, I think it's getting long enough to maybe warrant refactoring or splitting in a separate PR.

File: client/my-sites/plans/jetpack-plans/selector.tsx [hunk 4]
```diff
@@ -57,6 +58,11 @@ const SelectorPage: React.FC< SelectorPageProps > = ( {
 	const viewTrackerProps = siteId ? { site: siteSlug } : {};
 	const legacyPlan = planRecommendation ? planRecommendation[ 0 ] : null;
 
+	const [ isLoadingUpsellPageExperiment, experimentAssignment ] = useExperiment(
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/selector.tsx:61]: Not sure if it makes sense to call `useExperiment` in the product card component to avoid the need to pass the `isLoadingUpsellPageExperiment` prop through a few layers of the component tree?

Could go either way there.

File: client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts [hunk 5]
```diff
@@ -49,6 +56,24 @@ function buildCheckoutURL(
...
+				urlQueryArgs.checkoutBackUrl = window.location.href;
+			}
+
+			return addQueryArgs(
+				urlQueryArgs,
+				`${ rootUrl.replace( /\/$/, '' ) }/upsell/${ product }/${ siteSlug }`
+			);
+		}
+	}
+
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts]: Would it make sense to set this up as a separate function like `buildUpsellURL` since it's not returning a checkout URL? `showUpsellPage` could be used to conditionally call a function in `getPurchaseURLCallback`?

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+	const isFetchingPurchases = useSelector( isFetchingSitePurchases );
+	const purchases = useSelector( ( state ) => getSitePurchases( state, siteId ) );
+
+	const productItem = useMemo( () => slugToSelectorProduct( productSlug ), [ productSlug ] );
+	const upsellItem = useMemo(
+		() => ( upsellSlug ? slugToSelectorProduct( upsellSlug ) : null ),
+		[ upsellSlug ]
+	);
+	const productCheckoutURL = useMemo(
+		() => buildCheckoutURL( siteSlug, productSlug, urlQueryArgs ),
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/upsell/index.tsx:60]: Is there a reason not to use `getPurchaseURLCallback` here?

Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/upsell/index.tsx:60]: I suppose the `urlQueryArgs` here have something to do with this issue mentioned in the test plan:

> Note: clicking the close button in the checkout page should redirect you to the pricing page (passed in the checkoutBackUrl query param). This doesn't work as expected, and should be addressed in a separate PR.

I'm not sure what a fix for that looks like at the moment, is it too much additional code to handle as part of this PR? Not against splitting into a sub-branch at all :) Just looking to learn more about what's needed for a fix.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+		( upsellPriceObj?.discountedPrice || upsellPriceObj?.originalPrice ) -
+		( productPriceObj?.discountedPrice || productPriceObj?.originalPrice );
+	const isLoadingPrice = productPriceObj?.isFetching || upsellPriceObj?.isFetching;
+
+	useEffect( () => {
+		const dontShowUpsell =
+			! isLoadingUpsellPageExperiment && experimentAssignment?.variationName !== 'treatment';
+
+		if ( ( ! upsellSlug || isUpsellOwned || dontShowUpsell ) && productCheckoutURL ) {
+			page.redirect( productCheckoutURL );
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/upsell/index.tsx:114]: When I did the regression testing by accessing the upsell URL directly:

> Visit /plans/upsell/jetpack_backup_t1_yearly/:site: you should be redirected to the checkout page (you might briefly see the upsell page flash, but that's acceptable since we'll eventually keep or get rid of the page definitively)

If I tried to clear the cart from the checkout page, I was immediately redirected back to the checkout page. Not sure if this can cause that redirect unintentionally.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+					<ul>
+						{ features.map( ( { slug, text } ) => (
+							<li key={ slug }>
+								{ productNames[ slug ] || slug } -{ ' ' }
+								{ slug === productSlug ? translate( 'Already in your cart' ) : text }
+							</li>
+						) ) }
+					</ul>
+				) }
+				{ ! isLoadingPrice && priceDelta > 0 && (
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/upsell/index.tsx:157]: Is there a way to show a loading state here? (may be in a separate PR already)
When this loads the CTA buttons move down - I missed clicking on them a few times :)

**Review by jboland88 — COMMENTED** [review]
I think this looks pretty good overall. I left a few comments/questions to discuss.
I hit one redirect bug in the regression test that I mentioned inline. Otherwise, this tested well for me.

I'm not too familiar with how the pages/ routing works in Calypso, it may be good to get 👀 from another reviewer with some more knowledge about that. It looks sensible to me though.

File: client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts (hunk 5 above)
Comment (monsieur-z) [reply client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts]: Absolutely. I originally thought they would be more intertwined.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx (hunk 6 above)
Comment (monsieur-z) [reply client/my-sites/plans/jetpack-plans/upsell/index.tsx:60]: The checks in `getPurchaseURLCallback` are not relevant in this context, so better call `buildCheckoutURL` right away.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx (hunk 8 above)
Comment (monsieur-z) [reply client/my-sites/plans/jetpack-plans/upsell/index.tsx:157]: Yes, that's handled in #64329.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx (hunk 6 above)
Comment (monsieur-z) [reply client/my-sites/plans/jetpack-plans/upsell/index.tsx:60]: Intuitively, it seems that the close button in the checkout page is not taking into account the `checkoutBackUrl` query parameter. We'd need to confirm that's the actual issue.

This might soon not be relevant anymore: p1HpG7-gg7-p2.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx (hunk 7 above)
Comment (monsieur-z) [reply client/my-sites/plans/jetpack-plans/upsell/index.tsx:114]: Good catch, redirecting to the pricing page instead.

File: client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts [hunk 9]
```diff
@@ -105,5 +151,12 @@ export const getPurchaseURLCallback =
...
 
-		return buildCheckoutURL( siteSlug, product.productSlug, urlQueryArgs );
+		let url;
+
+		// Link to upsell page if upsell feature enabled
+		if ( showUpsellPage ) {
+			url = buildUpsellURL( siteSlug, product.productSlug, urlQueryArgs, rootUrl );
+		}
+
+		return url || buildCheckoutURL( siteSlug, product.productSlug, urlQueryArgs );
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts:161]: Nitpick: would it make sense to assign the default value to url as the variable is defined?

I see the use case for using `||` since `buildUpsellURL` can return `string` or `null`.

**Review by jboland88 — CHANGES_REQUESTED** [review]
I just went through the functional testing this time, I will return to take a look at the code changes. Noted a few things.

In the regression testing, this did not work for me:

> Visit /plans/upsell/jetpack_backup_t1_yearly/:site: you should be redirected to the checkout page (you might briefly see the upsell page flash, but that's acceptable since we'll eventually keep or get rid of the page definitively)

I was directed back to the pricing page at `plans/:site` for this step. 

While viewing the offer page:
> Refresh the page, and verify that you see the same information

Sometimes I would see a different price that appeared to be the difference in the original prices of the two products. When testing on the cloud pricing page, sometimes I would not see any price after a refresh.

When I remove the item from my cart after going through the upsell screen (not clicking the close button), I am redirected back to the upsell screen. It seems like I should end up back at `plans/:site`.  This seems to happen when starting at `plans/`, `plans/:site`, `/jetpack/connect/store` and `/jetpack/connect/plans/:site`
This may be related to the noted issue with the close button on the checkout page. Overall, it seems like an incorrect flow. I think [this file](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/hooks/use-valid-checkout-back-url.ts) may have something to do with it. `checkoutBackUrl` appears to be configured for a given set of hosts. Perhaps an additional check [in the leaveCheckout function](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/lib/leave-checkout.ts#L36-L43) could work around pages with `/upsell/` as part of the url structure?

> Visit /jetpack/connect/store

Something seems incorrect when starting from `/jetpack/connect/store` with no site specified. For the items that show the upsell screen, the first time that I click on the CTA to get the product, nothing happens (but the URL changes). When I click the CTA again, I see the upsell screen, but the CTAs from the upsell screen then take me to a site selection interface and not to checkout.

**Comment by monsieur-z** [comment]
> In the regression testing, this did not work for me:

> I was directed back to the pricing page at plans/:site for this step.

Yes, that's [now intended](https://github.com/Automattic/wp-calypso/pull/63828/#discussion_r892524252). I'll update the instructions too.

File: client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts (hunk 9 above)
Comment (monsieur-z) [reply client/my-sites/plans/jetpack-plans/get-purchase-url-callback.ts:161]: I'd say let's avoid calling a function when you don't need it.

**Comment by monsieur-z** [comment]
> Sometimes I would see a different price that appeared to be the difference in the original prices of the two products. When testing on the cloud pricing page, sometimes I would not see any price after a refresh.

Good catch! Should be fixed now.

> Something seems incorrect when starting from /jetpack/connect/store with no site specified.

This one too.

> When I remove the item from my cart after going through the upsell screen (not clicking the close button), I am redirected back to the upsell screen. It seems like I should end up back at plans/:site.

I'll address that in a separate PR, since updating checkout related functions might have unexpected implications. I don't think this is a blocker.

File: client/my-sites/plans/jetpack-plans/upsell/index.tsx (hunk 6 above)
Comment (jboland88) [reply client/my-sites/plans/jetpack-plans/upsell/index.tsx:60]: Resolving in favor of addressing in a separate PR 👍

**Review by jboland88 — APPROVED** [review]
This tested well with one small exception:

> Refresh the page, and verify that you see the same information

Occasionally, I would still see the price difference between the original prices of the two products after a refresh. It seemed inconsistent, but happened most readily when testing `plans/:site`. The price difference was always correct for me on the first load. I don't think this should block the whole PR, but would be good to see if we can resolve that.

As I mentioned before, I don't have as much knowledge on how Calypso handles routing, but the code changes seem sensible to me. 

With regards to handling redirection when items are removed from the cart:

> I'll address that in a separate PR, since updating checkout related functions might have unexpected implications. I don't think this is a blocker.

Sounds good, thanks.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7379784

Hi @monsieur-z, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Here’s a popular bundle for comprehensive site security:`
- `Save money with all Jetpack security products in one bundle:`
- `Upgrade your site security for only an additional{{asterisk/}}:`
- `Real-time cloud backups and one-click restores`
- `Block spam in comments and forms`
- `Nice choice, we added %(productName)s to your cart.`
- `Best value`

Thank you in advance!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7379784).
